editorial review of recent databus doc changes (#6199)
Particular/docs.particular.net: nservicebus/messaging/databus/index.md

## Transport message size limits

> The data bus should be used when it expected that message size may exceed the message size limit for the transport being used.

Reviewer: ```suggestion
The data bus should be used when it is expected that message size may exceed the message size limit for the transport being used.
```

## Alternatives

NOTE: A combination of these techniques may be used.

- Use a different transport or different tier (e.g. Azure Service Bus _Premium_ instead of _Standard_).

> - Message compression: this works well on text-based payloads like XML and JSON, or any payload (text or binary) that contains repetitive data.

Reviewer 2: ```suggestion
- Message compression: This works well on text-based payloads like XML and JSON, or any payload (text or binary) that contains repetitive data.
```

Author: I think it should stay as-is. It's arguable that the text after the colon is not a complete sentence since it relies on the "this" context from before the colon, and most style have it optional anyway for complete sentences. I would say it's better to be consistent with lists and partial sentences and start with lower case.

  - See the [message mutator example demonstrating message body compression](/samples/messagemutators/).
- Stream-based properties: The sample showing [handling large stream properties via pipeline](/samples/pipeline/stream-properties/) demonstrates a purely stream-based approach (rather than loading the full payload into memory) using the NServiceBus pipeline.
- A more efficient serializer: For example, a binary serializer. Most serializers can be added with a few lines of code.
   - Several [serializers are maintained by the community](/nservicebus/community/#serializers).

> - Attachments: the community maintained [NServiceBus.Attachments](/nservicebus/community/#nservicebus-attachments) package may be used for unbounded binary payloads.

Author: see https://github.com/Particular/docs.particular.net/pull/6199#discussion_r1335520372

>   - Read on demand: attachments are only retrieved when read by a consumer.

Reviewer 2: ```suggestion
  - Read on demand: Attachments are only retrieved when read by a consumer. Moreover, it processes all data items via an IAsyncEnumerable
```

Author: OK, here we have two sentences so a capital is warranted. I'll consider the list as a whole.

> - Attachments: The community maintained [NServiceBus.Attachments](/nservicebus/community/#nservicebus-attachments) package may be used for unbounded binary payloads. Although it has functionality similar to the [DataBus](https://docs.particular.net/nservicebus/messaging/databus/), it has the following advantages.

Author: ```suggestion
- Attachments: The community maintained [NServiceBus.Attachments](/nservicebus/community/#nservicebus-attachments) package may be used for unbounded binary payloads. Although it has functionality similar to the [DataBus](nservicebus/messaging/databus/), it has the following advantages.
```

>   - Reduced memory usage: the package avoids the use of a base64 serializer which results in a significant reduction in memory usage.

Reviewer 2: ```suggestion
  - Reduced memory usage: The package avoids the use of a base64 serializer which results in a significant reduction in memory usage.
```

> - Attachments: The community maintained [NServiceBus.Attachments](/nservicebus/community/#nservicebus-attachments) package may be used for unbounded binary payloads. Although it has functionality similar to the [DataBus](nservicebus/messaging/databus/), it has the following advantages.
>   - Read on demand: Attachments are only retrieved when read by a consumer. Moreover, it processes all data items via an IAsyncEnumerable
>   - Reduced memory usage: The package avoids the use of a base64 serializer which results in a significant reduction in memory usage.

Author: @soujay

```suggestion
- Attachments: The community maintained [NServiceBus.Attachments](/nservicebus/community/#nservicebus-attachments) package may be used for unbounded binary payloads. The package is similar to the [Data Bus](nservicebus/messaging/databus/) but has some differences:
  - Read on demand: Attachments are only retrieved when read by a consumer.
  - Async enumeration: The package also supports processing all data items using an `IAsyncEnumerable`.
  - Reduced memory usage: The package avoids the use of a base64 serializer which results in a significant reduction in memory usage.
```

Note that these are not necessarily _advantages_. They are beneficial in specific circumstances.

Reviewer: @soujay @adamralph how about:

```suggestion
- Attachments: The community maintained [NServiceBus.Attachments](/nservicebus/community/#nservicebus-attachments) package may be used for unbounded binary payloads. The package is similar to the Data Bus but has some differences:
  - Read on demand: Attachments are only retrieved when read by a consumer.
  - Async enumeration: The package supports processing all data items using an `IAsyncEnumerable`.
  - Reduced memory usage: The package avoids the use of a base64 serializer which results in a significant reduction in memory usage.
  - Better suited for [binary large objects (BLOBs](https://en.wikipedia.org/wiki/Binary_large_object)
```

Note that these are not necessarily _advantages_. They are beneficial in specific circumstances.

Also note that the URL would be a link back to the page being read, so I've removed it.

Author: @ramonsmits This is a list of differences. What is the difference that makes it better suited for BLOBs?

Reviewer: > @ramonsmits This is a list of differences. What is the difference that makes it better suited for BLOBs?

Added that to my previous edit

Author: Thanks @ramonsmits. I'll commit my suggestion first and then continue this thread so it's easier to see the change you're proposing.

Author: continued in https://github.com/Particular/docs.particular.net/pull/6199/files#r1339836138

  - See the [message mutator example demonstrating message body compression](/samples/messagemutators/).
- Stream-based properties: The sample showing [handling large stream properties via pipeline](/samples/pipeline/stream-properties/) demonstrates a purely stream-based approach (rather than loading the full payload into memory) using the NServiceBus pipeline.
- A more efficient serializer: For example, a binary serializer. Most serializers can be added with a few lines of code.
   - Several [serializers are maintained by the community](/nservicebus/community/#serializers).
- Attachments: The community maintained [NServiceBus.Attachments](/nservicebus/community/#nservicebus-attachments) package may be used for unbounded binary payloads. The package is similar to the Data Bus but has some differences:
  - Read on demand: Attachments are only retrieved when read by a consumer.
  - Async enumeration: The package supports processing all data items using an `IAsyncEnumerable`.
  - Reduced memory usage: The package avoids the use of a base64 serializer which results in a significant reduction in memory usage.

> 

Author: @ramonsmits this is your suggestion from https://github.com/Particular/docs.particular.net/pull/6199#discussion_r1336862452 —

```suggestion
  - Better suited for [binary large objects (BLOBs](https://en.wikipedia.org/wiki/Binary_large_object) as it does not have (de)serialization overhead.

```

I'm not sure I understand and nor will a potential reader. This difference to the databus is [not listed in the package README](https://github.com/NServiceBusExtensions/NServiceBus.Attachments#compared-to-the-databus). Can you please explain more about why this package is better suited for BLOBS?

From the README I see:

> With the DataBus the only interaction is via byte arrays. NServiceBus.Attachments supports reading and writing using streams, byte arrays, or string.

How does this make it better for BLOBs?

Reviewer: The databus stores byte arrays (as everything does...) but handles all this in memory and (de)serializes properties to/from byte arrays which is different with NServiceBus.Attachments

If you actually have BLOB's like large files NServiceBus.Attachments is by far the better option.

Author: > which is different with NServiceBus.Attachments

How is it different? What does NServiceBus.Attachments do?

> If you actually have BLOB's like large files NServiceBus.Attachments is by far the better option.

Sorry, maybe I'm being slow, but I still don't understand why.

Reviewer: > Sorry, maybe I'm being slow, but I still don't understand why.

Well, maybe then read documentation on how streams work?

Reviewer: @adamralph WHen using streams and you are purely processing binary BLOB's like PDF's, image, video, backup files, logs,e tc

a. Things don't need to be fully loaded in memory. These BLOBs are likely being read from (relatively slow) file/network streams anyway to just passing data directly without requiring to load these in memory.
b. Don't require (de)serialization by very inefficiently loading binary data, then serializing that to Base64 for storage.

Author: @ramonsmits thanks, I think I get it now. This seems to tie in with the previous bullet, so I'll make a suggestion on that.

Author: _continued in https://github.com/Particular/docs.particular.net/pull/6199/files#r1372717404_

- Use a different transport or different tier (e.g. Azure Service Bus _Premium_ instead of _Standard_).
- Use message body compression, which works well on text-based payloads like XML and JSON or any payload (text or binary) that contains repetitive data.
  - The [message mutator sample](/samples/messagemutators/) demonstrates message body compression.
- Use stream-based properties.
  - The sample showing how to [handle large stream properties via the pipeline](/samples/pipeline/stream-properties/) demonstrates a purely stream-based approach (instead of loading full payloads into memory).
- Use a more efficient serializer, such as a binary serializer.
  - A custom serializer can usually be implemented in only a few lines of code.
  - Some binary [serializers are maintained by the community](/nservicebus/community/#serializers).
- Use [NServiceBus.Attachments](/nservicebus/community/#nservicebus-attachments) for unbounded binary payloads. The package is similar to the Data Bus but has some differences:
  - Read on demand: Attachments are only retrieved when read by a consumer.
  - Async enumeration: The package supports processing all data items using an `IAsyncEnumerable`.

>   - Reduced memory usage: The package avoids the use of a base64 serializer which results in a significant reduction in memory usage.

Author: _continued from https://github.com/Particular/docs.particular.net/pull/6199/files#r1372702758_

```suggestion
  - No base64 serialization: This results in a significant reduction in memory usage.
  - Direct stream access: This makes the package more suitable for [binary large objects (BLOBs](https://en.wikipedia.org/wiki/Binary_large_object) since stream contents do not necessarily have to be loaded into memory before storing them or when retrieving them.
```

@ramonsmits where is base64 serialization used? I.e. a significant reduction in memory usage compared to what?

Reviewer: @adamralph In the serializers when dealing with byte arrays. ALL that happens in memory and if you are dealing with binary streams to begin with then there is no need for all of this overhead.

Author: ```suggestion
  - No serialization: The serializer is not used, which may result in a significant reduction in memory usage.
  - Direct stream access: This makes the package more suitable for [binary large objects (BLOBs](https://en.wikipedia.org/wiki/Binary_large_object) since stream contents do not necessarily have to be loaded into memory before storing them or when retrieving them.
```
